### PR TITLE
Onboarding task8

### PR DESCRIPTION
**Pull request by thalyscarrara**
## Description

implement coroutines in codelab

## Related Issues

[task08-onboarding](https://github.com/profusion/quickstart-android-jetpack-graphql/issues/210)

## Visual reference

https://user-images.githubusercontent.com/53837410/193031003-7c04023e-c438-4a3c-b929-af77574e5603.mp4

### CODE REVIEW

File: coroutines-codelab/start/src/main/java/com/example/android/kotlincoroutines/main/MainViewModel.kt [hunk 1]
```diff
@@ -102,10 +105,15 @@ class MainViewModel(private val repository: TitleRepository) : ViewModel() {
...
         // TODO: Convert updateTaps to use coroutines
-        tapCount++
-        BACKGROUND.submit {
-            Thread.sleep(1_000)
-            _taps.postValue("${tapCount} taps")
+//        tapCount++
+//        BACKGROUND.submit {
+//            Thread.sleep(1_000)
+//            _taps.postValue("${tapCount} taps")
+//        }
```
Comment (mari1912) [inline coroutines-codelab/start/src/main/java/com/example/android/kotlincoroutines/main/MainViewModel.kt]: remove it

**Review by mari1912 — APPROVED** [review]
Good work! Please, just remove the parts of code that you commented instead of deleting

File: coroutines-codelab/start/src/main/java/com/example/android/kotlincoroutines/main/MainViewModel.kt [hunk 2]
```diff
@@ -123,16 +123,26 @@ class MainViewModel(private val repository: TitleRepository) : ViewModel() {
...
+//        repository.refreshTitleWithCallbacks(object : TitleRefreshCallback {
+//            override fun onCompleted() {
+//                _spinner.postValue(false)
+//            }
+//
+//            override fun onError(cause: Throwable) {
+//                _snackBar.postValue(cause.message)
+//                _spinner.postValue(false)
+//            }
+//        })
```
Comment (joaobertacchi) [inline coroutines-codelab/start/src/main/java/com/example/android/kotlincoroutines/main/MainViewModel.kt]: ?

File: coroutines-codelab/start/src/main/java/com/example/android/kotlincoroutines/main/TitleRepository.kt [hunk 3]
```diff
@@ -41,8 +43,21 @@ class TitleRepository(val network: MainNetwork, val titleDao: TitleDao) {
      */
     val title: LiveData<String?> = titleDao.titleLiveData.map { it?.title }
 
-
     // TODO: Add coroutines-based `fun refreshTitle` here
```
Comment (joaobertacchi) [inline coroutines-codelab/start/src/main/java/com/example/android/kotlincoroutines/main/TitleRepository.kt]: please remove the TODO

File: coroutines-codelab/start/src/debug/java/com/example/android/kotlincoroutines/fakes/TestingFakes.kt [hunk 4]
```diff
@@ -101,7 +101,7 @@ class MainNetworkFake(var result: String) : MainNetwork {
 class MainNetworkCompletableFake() : MainNetwork {
     private var completable = CompletableDeferred<String>()
 
-    override fun fetchNextTitle() = MakeCompilerHappyForStarterCode() // TODO: replace with `completable.await()`
+    override suspend fun fetchNextTitle() = completable.await() // TODO: replace with `completable.await()`
```
Comment (joaobertacchi) [inline coroutines-codelab/start/src/debug/java/com/example/android/kotlincoroutines/fakes/TestingFakes.kt]: remove TODO

**Review by joaobertacchi — COMMENTED** [review]
Your 2nd commit (actually all the commits) must start with a verb in the infinitive: `feat: make...`.

**Review by joaobertacchi — COMMENTED** [review]
Pre-approving. Please update your commit titles to follow commit rules: https://github.com/profusion/quickstart-android-jetpack-graphql/blob/master/docs/CONTRIBUTING.md#description

File: coroutines-codelab/start/src/main/java/com/example/android/kotlincoroutines/main/TitleRepository.kt [hunk 5]
```diff
@@ -41,8 +43,20 @@ class TitleRepository(val network: MainNetwork, val titleDao: TitleDao) {
...
-    // TODO: Add coroutines-based `fun refreshTitle` here
+    suspend fun refreshTitle() {
+        withContext(Dispatchers.IO) {
+            val result = try {
+                network.fetchNextTitle().execute()
+            } catch (cause: Throwable) {
+                throw TitleRefreshError("Unable to refresh title", cause)
+            }
+            if (result.isSuccessful) {
+                titleDao.insertTitle(Title(result.body()!!))
```
Comment (paulohc) [inline coroutines-codelab/start/src/main/java/com/example/android/kotlincoroutines/main/TitleRepository.kt]: Avoid to use `!!`. Do a null check instead

**Comment by paulohc** [comment]
I received some `Unable to refresh title` errors. Do you know why this is happening?

**Review by paulohc — APPROVED** [review]

**Review by joaobertacchi — APPROVED** [review]

**Pull request merged**